### PR TITLE
lexer change: Allow emphasis in CJK text without spaces

### PR DESCRIPTION
**Pull request by peng1999**
Chinese and Japanese do not use spaces in normal texts. So currently you always need to use function instead of markups for emphasis, which is annoying. This PR fixes that.

Related: #635 ([comment](https://github.com/typst/typst/issues/635#issuecomment-1790221701))

### CODE REVIEW

File: crates/typst-syntax/src/lexer.rs [hunk 1]
```diff
@@ -343,7 +344,15 @@ impl Lexer<'_> {
     }
 
     fn in_word(&self) -> bool {
-        let alphanum = |c: Option<char>| c.map_or(false, |c| c.is_alphanumeric());
+        let alphanum = |c: Option<char>| {
```
Comment (laurmaedje) [inline crates/typst-syntax/src/lexer.rs:347]: The name `alphanum` is a bit misleading here now, right? It should probably be changed.

**Review by laurmaedje — COMMENTED** [review]
Do we really need two large bold fonts in the repo for testing or would one suffice? I assume this is to test different scripts?

File: crates/typst-syntax/src/lexer.rs (hunk 1 above)
Comment (peng1999) [reply crates/typst-syntax/src/lexer.rs:347]: What do you think is a good name? Do you like the name `non_cjk_alphanum`?

**Comment by peng1999** [comment]
> Do we really need two large bold fonts in the repo for testing or would one suffice?

Actually in the test I added in this PR, only the sc-bold variant font is needed. But the irrelevant test `lang-with-region.typ` will not look right after I added the font:

lang-with-region

The outlines should be bold, but previously it is not I nobody noticed that because there is no bold CJK font at all. 

So, in summary:

- `emphasis.typ`: needs one sc-bold font to look right.
- `lang-with-region.typ`
   - needs both of sc-bold and tc-bold to really be right.
   - ... or get rid of both to _pretend_ to be right.
   - ... or leave one bold font, to be obviously looks wrong.

File: crates/typst-syntax/src/lexer.rs (hunk 1 above)
Comment (laurmaedje) [reply crates/typst-syntax/src/lexer.rs:347]: Either something descriptive like that or something like `wordy` in line with `in_word`.

**Comment by laurmaedje** [comment]
I guess we can keep both fonts.

**Comment by laurmaedje** [comment]
Thank you, I think this will really help CJK users!

**Comment by jiaojiaodubai** [comment]
> > Do we really need two large bold fonts in the repo for testing or would one suffice?
> 
> The outlines should be bold, but previously it is not I nobody noticed that because there is no bold CJK font at all.
> 

I would like to talk something I know about it.

Traditionally, **there is no concept of "bold" in Chinese publishing technology**. When emphasizing, we usually replace a font type, such as regular script or imitation Song script. Therefore, Chinese font files often do not contain multiple word weights. 
image

However, the emergence of Microsoft Word allows people to use 'Fake Bold' to bold any character for emphasis, and many people have become accustomed to using bolder characters for emphasis. In this trend, some new fonts (such as Source Han) have also begun to release multiple font weights.

It should be noted that **emphasizing in bold is still not a common practice in official publications** and is usually only done in (informal) electronic documents.

In the Chinese macro set of LaTeX, the default behavior of the `\emph{}` command is consistent with the practice of traditional publications, where emphasis is expressed in regular script instead of bold. I think we also need to distinguish between "emphasis" and "bold" in the Chinese context.

**Comment by laurmaedje** [comment]
We already have a distinction between normal emphasis (`emph` and \_), strong emphasis (`strong` and \*), italic (`text(style: "italic")`), and bold (`text(weight: "bold")`).

This means we _could_ change the default behaviour of emph/strong based on the current text language. And users can of course also do it themselves with a show rule.

**Comment by GES233** [comment]
There's may something wrong in `emphasis.typ`. It's  _斜体_（Italic） not 楷体（Kai Font）.

**Comment by laurmaedje** [comment]
This PR only ensured that emphasis is parsable in CJK text. Special handling for Kai font instead of italics is not yet implemented.

**Comment by tats-u** [comment]
The Scripts-based approach doesn't cover some characters:

- https://github.com/typst/typst/pull/2648
- https://github.com/prettier/prettier/pull/16832

**Comment by tats-u** [comment]
Fullwidth alphanumerics are not covered by this approach.

**Pull request merged**